### PR TITLE
Add model parameter to playbook gen/exp APIs

### PR DESCRIPTION
**Pull request by TamiTakamiya**
Jira Issue: <https://issues.redhat.com/browse/AAP-32137>

## Description

Add model parameter to playbook gen/exp APIs.

### Steps to test
1. Pull down the PR
2. Run Unit tests

### Scenarios tested

Unit test cases are included.

## Production deployment

- [X] This code change is ready for production on its own
- [ ] This code change requires the following considerations before going to production:

### CODE REVIEW

File: ansible_ai_connect/ai/api/model_client/tests/test_wca_client.py [hunk 1]
```diff
@@ -385,6 +405,26 @@ def test_playbook_exp_error(self):
...
+        def get_my_model_id(user, organization_id, model_id):
+            self.assertEqual(model_id, "mymodel")
+            self.assertion_count += 1
+            return model_id
+
+        model_client.get_model_id = get_my_model_id
+
+        model_client.explain_playbook(request, content="Some playbook", model_id="mymodel")
+        self.assertGreater(self.assertion_count, 0)
+        self.assertion_count = 0
```
Comment (manstis) [inline ansible_ai_connect/ai/api/model_client/tests/test_wca_client.py:426]: Is this needed?

File: ansible_ai_connect/ai/api/model_client/tests/test_dummy_client.py [hunk 2]
```diff
@@ -70,8 +70,23 @@ def test_generate_playbook_with_outline(self):
         self.assertTrue(isinstance(outline, str))
         self.assertTrue(outline)
 
+    def test_generate_playbook_with_model_id(self):
```
Comment (manstis) [inline ansible_ai_connect/ai/api/model_client/tests/test_dummy_client.py:73]: I don't really understand what this is testing.

It appears to test that the type of the return parameters is correct; but doesn't really test `mymodel` was used.

File: ansible_ai_connect/ai/api/model_client/tests/test_dummy_client.py [hunk 3]
```diff
@@ -70,8 +70,23 @@ def test_generate_playbook_with_outline(self):
...
+        self.assertTrue(isinstance(outline, str))
+        self.assertTrue(outline)
+
     def test_explain_playbook(self):
         client = DummyClient(inference_url="https://ibm.com")
         explanation = client.explain_playbook(None, "ëoo")
         self.assertTrue(isinstance(explanation, str))
         self.assertTrue(explanation)
+
+    def test_explain_playbook_with_model_id(self):
```
Comment (manstis) [inline ansible_ai_connect/ai/api/model_client/tests/test_dummy_client.py:88]: Likewise. This doesn't really seem to test that `mymodel` was used.

File: ansible_ai_connect/ai/api/model_client/tests/test_wca_client.py [hunk 4]
```diff
@@ -360,6 +360,26 @@ def test_playbook_gen_error(self):
...
+            self.assertEqual(model_id, "mymodel")
+            self.assertion_count += 1
+            return model_id
+
+        model_client.get_model_id = get_my_model_id
+
+        model_client.generate_playbook(
+            request, text="Install Wordpress", create_outline=True, model_id="mymodel"
+        )
+        self.assertGreater(self.assertion_count, 0)
```
Comment (manstis) [inline ansible_ai_connect/ai/api/model_client/tests/test_wca_client.py:381]: This is really only asserting that your mock `get_my_model_id` was called.

Would it not be better to check `WCAClient.get_model_id` was called?

File: ansible_ai_connect/ai/api/model_client/tests/test_wca_client.py [hunk 5]
```diff
@@ -385,6 +405,26 @@ def test_playbook_exp_error(self):
...
+
+        def get_my_model_id(user, organization_id, model_id):
+            self.assertEqual(model_id, "mymodel")
+            self.assertion_count += 1
+            return model_id
+
+        model_client.get_model_id = get_my_model_id
+
+        model_client.explain_playbook(request, content="Some playbook", model_id="mymodel")
+        self.assertGreater(self.assertion_count, 0)
```
Comment (manstis) [inline ansible_ai_connect/ai/api/model_client/tests/test_wca_client.py:425]: This is really only asserting that your mock `get_my_model_id` was called.

Would it not be better to check `WCAClient.get_model_id` was called?

File: ansible_ai_connect/ai/api/tests/test_views.py [hunk 6]
```diff
@@ -2416,6 +2422,34 @@ def test_ok(self):
...
+          ansible.builtin.dnf:
+            name: nginx
+            state: present
+    """,
+            "explanationId": explanation_id,
+            "ansibleExtensionVersion": "24.4.0",
+            "model": model,
+        }
+        self.client.force_authenticate(user=self.user)
+        with self.assertLogs(logger="root", level="DEBUG") as log:
```
Comment (manstis) [inline ansible_ai_connect/ai/api/tests/test_views.py:2444]: Should we not be asserting somewhere that `mymodel` was used!?!

File: ansible_ai_connect/ai/api/tests/test_views.py [hunk 7]
```diff
@@ -2647,10 +2683,11 @@ def assert_test(
         with self.assertLogs(logger="root", level="DEBUG") as log:
             r = self.client.post(reverse("explanations"), self.payload, format="json")
             self.assertEqual(r.status_code, expected_status_code)
-            self.assert_error_detail(
-                r, expected_exception().default_code, expected_exception().default_detail
-            )
-            self.assertInLog(expected_log_message, log)
+            if expected_status_code != HTTPStatus.OK:
```
Comment (manstis) [inline ansible_ai_connect/ai/api/tests/test_views.py:2686]: Would it not be better to check if `excepted_exception() !== None`?

[This](https://github.com/ansible/ansible-ai-connect-service/blob/c5462a42a7e20e403ff6a07cec82fbf00ef9a6f9/ansible_ai_connect/ai/api/tests/test_views.py#L2849-L2854) example looks like it'd break with your change.

File: ansible_ai_connect/ai/api/tests/test_views.py [hunk 8]
```diff
@@ -2792,6 +2829,31 @@ def test_wca_instance_deleted(self):
...
+            "explanationId": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
+        }
+
+        self.assert_test(
+            model_client,
+            HTTPStatus.OK,
+            None,
+            None,
+        )
+        self.assertGreater(self.assertion_count, 0)
```
Comment (manstis) [inline ansible_ai_connect/ai/api/tests/test_views.py:2855]: This only asserts that your mock `get_model_id` was called.

We could have done the following and check an exception was raised.
```
model_client.get_model_id = Mock(side_effect=Exception)
```
It doesn't test that the _real_ method was called.

File: ansible_ai_connect/ai/api/tests/test_views.py [hunk 9]
```diff
@@ -2848,6 +2910,23 @@ def test_ok(self):
...
+            "generationId": generation_id,
+            "ansibleExtensionVersion": "24.4.0",
+            "model": model,
+        }
+        self.client.force_authenticate(user=self.user)
+        r = self.client.post(reverse("generations"), payload, format="json")
+        self.assertEqual(r.status_code, HTTPStatus.OK)
+        self.assertIsNotNone(r.data["playbook"])
+        self.assertEqual(r.data["format"], "plaintext")
+        self.assertEqual(r.data["generationId"], generation_id)
```
Comment (manstis) [inline ansible_ai_connect/ai/api/tests/test_views.py:2928]: We're not asserting that `mymodel` was used.

File: ansible_ai_connect/ai/api/tests/test_views.py [hunk 10]
```diff
@@ -3116,10 +3197,11 @@ def assert_test(
         with self.assertLogs(logger="root", level="DEBUG") as log:
             r = self.client.post(reverse("generations"), self.payload, format="json")
             self.assertEqual(r.status_code, expected_status_code)
-            self.assert_error_detail(
-                r, expected_exception().default_code, expected_exception().default_detail
-            )
-            self.assertInLog(expected_log_message, log)
+            if expected_status_code != 200:
```
Comment (manstis) [inline ansible_ai_connect/ai/api/tests/test_views.py:3200]: IMO checking if `expected_exception() !== None` would be better.

File: ansible_ai_connect/ai/api/tests/test_views.py [hunk 11]
```diff
@@ -3261,6 +3343,28 @@ def test_wca_instance_deleted(self):
...
+
+        model_client.get_model_id = get_model_id
+        model_client.generate_playbook = lambda *args: ("playbook", "outline")
+        self.assert_test(
+            model_client,
+            HTTPStatus.OK,
+            None,
+            None,
+        )
+        self.assertGreater(self.assertion_count, 0)
```
Comment (manstis) [inline ansible_ai_connect/ai/api/tests/test_views.py:3366]: Same comment about not really checking the real method did what we expected. Sorry.

**Review by manstis — COMMENTED** [review]
The _code changes_ look OK.

I am however concerned the new tests don't really check the code changes; but simply check your mocks worked as expected. Sorry.

If @goneri or @jameswnl think differently then please ignore my review.

File: ansible_ai_connect/ai/api/model_client/tests/test_wca_client.py (hunk 4 above)
Comment (TamiTakamiya) [reply ansible_ai_connect/ai/api/model_client/tests/test_wca_client.py:381]: I wanted to assert `model_id` was passed to `get_model_id` when WCA playbook gen is used and thought if the mocked `get_model_id` received the given `model_id` value, it was asserted even though it is not the real `WCAClient.get_model_id`.

> Would it not be better to check WCAClient.get_model_id was called?

I thought I might add a DEBUG trace to `WCAClient.get_model_id` and check the log later, but did not take that option.  If it looks more adequate or there is a better different approach, I am fully open for the change.

File: ansible_ai_connect/ai/api/model_client/tests/test_wca_client.py (hunk 5 above)
Comment (TamiTakamiya) [reply ansible_ai_connect/ai/api/model_client/tests/test_wca_client.py:425]: Same as above.

File: ansible_ai_connect/ai/api/tests/test_views.py (hunk 8 above)
Comment (TamiTakamiya) [reply ansible_ai_connect/ai/api/tests/test_views.py:2855]: Ok, I see. I will try to implement that. Thanks.

**Review by manstis — APPROVED** [review]
Approving on the basis of the _code changes_.

I understand that the tests will be revisited in due course.

**Comment by TamiTakamiya** [comment]
@manstis Thanks!

File: ansible_ai_connect/ai/api/model_client/tests/test_dummy_client.py (hunk 2 above)
Comment (TamiTakamiya) [reply ansible_ai_connect/ai/api/model_client/tests/test_dummy_client.py:73]: It is just for making sure that `model_id` is accepted as a valid argument of the `generate_playbook` method of `DummyClient`

File: ansible_ai_connect/ai/api/model_client/tests/test_dummy_client.py (hunk 3 above)
Comment (TamiTakamiya) [reply ansible_ai_connect/ai/api/model_client/tests/test_dummy_client.py:88]: Same as above. It is just for making sure that `model_id` is accepted as a valid argument of the `explain_playbook` method of `DummyClient`

File: ansible_ai_connect/ai/api/tests/test_views.py (hunk 6 above)
Comment (TamiTakamiya) [reply ansible_ai_connect/ai/api/tests/test_views.py:2444]: Yes. We should be able to check the `modelName` value in segment events....

File: ansible_ai_connect/ai/api/tests/test_views.py (hunk 9 above)
Comment (TamiTakamiya) [reply ansible_ai_connect/ai/api/tests/test_views.py:2928]: Will check using a segment event...

**Pull request merged**